utilities/src/with_std/dynamic_events.rs

```diff
@@ -0,0 +1,419 @@
+pub struct EventDecoder {
+	events_type_id: u32,
+	types: PortableRegistry,
+	errors: NamedVariantLookup<ErrorInfo>,
+}
+
+impl EventDecoder {
+	/// Create a new event decoder.
+	///
+	/// The arguments to this constructor can be extracted from substrate metadata.
+	pub fn new(mut types: PortableRegistry, errors_type_id: u32) -> Option<Self> {
+		let errors = match types.resolve(errors_type_id).unwrap().type_def.clone() {
+			scale_info::TypeDef::Variant(runtime_error_type) => runtime_error_type
+				.variants
+				.into_iter()
+				.map(|pallet_error_type| {
+					(
+						pallet_error_type.index,
+						(pallet_error_type.name, {
+							let type_id = pallet_error_type
+								.fields
+								.first()
+								.expect("error variant has exactly one field")
+								.ty
+								.id;
+							match types.resolve(type_id).unwrap().type_def.clone() {
+								scale_info::TypeDef::Variant(pallet_errors) => pallet_errors
+									.variants
+									.into_iter()
+									.map(|error_variant| {
+										(
+											error_variant.index,
+											(error_variant.name, error_variant.docs),
+										)
+									})
+									.collect::<ErrorInfo>(),
+								_ => panic!("Inner error type is not an Enum"),
+							}
+						}),
+					)
+				})
+				.collect::<NamedVariantLookup<_>>(),
+			_ => panic!("Outer error type is not an Enum"),
+		};
+
+		let events_type_id = types
+			.types
+			.iter()
+			.find(|t| t.ty.path.segments.join("::") == SYSTEM_EVENTS_TYPE_NAME)
+			.map(|t| t.id)
+			.expect("System events type must exist in the metadata");
+
+		let events_type_id = *types
+			.retain(|id| events_type_id == id)
+			.get(&events_type_id)
+			.expect("outer event enum must exist in the metadata");
+
+		debug_assert_eq!(
+			types
+				.resolve(events_type_id)
+				.expect("events_type_id found in types above, so it must exist")
+				.path
+				.segments
+				.join("::"),
+			SYSTEM_EVENTS_TYPE_NAME
+		);
+
+		log::debug!("EventDecoder error lookup: {:?}", errors);
+
+		Some(Self { events_type_id, types, errors })
+	}
+
+	/// Decode a raw buffer of bytes into a list of [DyanmicEventRecord]s.
+	///
+	/// The input should be the raw scale encoded bytes queried from `System.Events` storage.
+	pub fn decode_events(
+		&self,
+		events_data: Vec<u8>,
+	) -> Result<Vec<DynamicEventRecord>, DecodeError> {
+		let events_data_cursor = &mut &events_data[..];
+
+		// The data represents a Vec<EventRecord<T, H>>, so it's prefixed with a Compact<u32>
+		// denoting the length of the Vec (ie. the number of events).
+		let event_count = Compact::<u32>::decode(events_data_cursor)
+			.expect("Failed to decode CompactLen")
+			.0;
+
+		let initial_len = events_data_cursor.len();
+		let decoded = (0..event_count)
+			.map(|_| {
+				ScaleDecodedToJson::decode_as_type(
+					events_data_cursor,
+					self.events_type_id,
+					&self.types,
+				)
+				.and_then(|decoded| {
+					let mut json: JsonValue = decoded.into();
+					if let Some(dispatch_error) =
```

Comment:
little comment that we're converting the error to a more useful/readable format here